trace/kafka/kafka.go

```diff
@@ -61,9 +83,11 @@ func (ap *AsyncProducer) Send(ctx context.Context, msg *Message) error {
 		opentracing.Tag{Key: "topic", Value: msg.topic})
 	pm, err := ap.createProducerMessage(ctx, msg, sp)
 	if err != nil {
+		messageStatusCountInc(messageCreationErrors, msg.topic)
```

Comment:
Would it make sense to catch the metrics outside the main logic ?
e.g. try to decorate the parent calls with a metrics layer.
So that we keep the metrics logic decoupled.
(see similar comment above for the consumer part)
That way we can also test the metrics better.

Reply:
I've stirred up a small example, decorating the createProducerMessage (cpm) function.    
Does it look something like what you had in mind?

```go
type cpm func(context.Context, *Message, opentracing.Span) (*sarama.ProducerMessage, error)

func (ap *AsyncProducer) measure(f cpm) cpm {
	return func(ctx context.Context, msg *Message, sp opentracing.Span) (*sarama.ProducerMessage, error) {
		pm, err := ap.createProducerMessage(ctx, msg, sp)
		if err != nil {
			messageStatusCountInc(messageCreationErrors, msg.topic)
			return pm, err
		}
		messageStatusCountInc(messageSent, msg.topic)
		return pm, err
	}
}
```

```go
// And then, instead of 
	pm, err := ap.createProducerMessage(ctx, msg, sp)
	if err != nil {
		messageStatusCountInc(messageCreationErrors, msg.topic)
		trace.SpanError(sp)
		return err
	}
	messageStatusCountInc(messageSent, msg.topic)
	ap.prod.Input() <- pm
	trace.SpanSuccess(sp)

// One would do
	pm, err := ap.measure(ap.createProducerMessage)(ctx, msg, sp)
	if err != nil {
		trace.SpanError(sp)
		return err
	}
	ap.prod.Input() <- pm
	trace.SpanSuccess(sp)
	return nil
```

I don't have a strong opinion on this, even though it seems more complicated.    
The only thing that I'm not sure it's okay, is that then we would have two actions; metrics and tracing, one of which would be decoupled to the decorator,  and the other will remain in the parent method.

Reply:
I think this looks cleaner, although a bit more complicated.
We might want to integrate the tracing inside this method as well , maybe ?
But happy to approve whatever you think makes more sense !

Reply:
it looks more complicated and i believe we should try to avoid it.